saleor/graphql/product/tests/mutations/test_product_variant_bulk_update.py

```diff
@@ -136,6 +147,90 @@ def test_product_variant_bulk_update(
         assert rule.variants_dirty
 
 
+@pytest.mark.parametrize(
+    "error_policy",
+    [ErrorPolicyEnum.REJECT_FAILED_ROWS.name, ErrorPolicyEnum.IGNORE_FAILED.name],
+)
+@patch(
+    "saleor.graphql.product.bulk_mutations."
+    "product_variant_bulk_update.get_webhooks_for_event"
+)
+@patch.object(ProductVariantBulkUpdate, "save_variants")
+def test_product_variant_bulk_create_stock_thread_race(
+    mocked_save,
+    mocked_get_webhooks_for_event,
+    error_policy,
+    staff_api_client,
+    variant_with_many_stocks,
+    warehouse,
+    permission_manage_products,
+    any_webhook,
+    settings,
+):
+    # given
+    mocked_get_webhooks_for_event.return_value = [any_webhook]
+    settings.PLUGINS = ["saleor.plugins.webhook.plugin.WebhookPlugin"]
+
+    variant = variant_with_many_stocks
+    new_stock_quantity_created_before_save = 999
+
+    def add_stock_before_save(*args, **kwargs):
+        stock = Stock(
+            warehouse=warehouse,
+            quantity=new_stock_quantity_created_before_save,
+            product_variant=variant,
+        )
+        stock.save()
+
+    mocked_save.side_effect = add_stock_before_save
+
+    product_id = graphene.Node.to_global_id("Product", variant.product_id)
+    variant_id = graphene.Node.to_global_id("ProductVariant", variant.pk)
+    stocks = variant.stocks.all()
+    assert len(stocks) == 2
+    new_stock_quantity = 100
+
+    variants = [
+        {
+            "id": variant_id,
+            "stocks": {
+                "create": [
+                    {
+                        "quantity": new_stock_quantity,
+                        "warehouse": graphene.Node.to_global_id(
+                            "Warehouse", warehouse.pk
+                        ),
+                    },
+                ],
+            },
+        }
+    ]
+
+    variables = {
+        "productId": product_id,
+        "variants": variants,
+        "errorPolicy": error_policy,
+    }
+
+    # when
+    staff_api_client.user.user_permissions.add(permission_manage_products)
+    response = staff_api_client.post_graphql(
+        PRODUCT_VARIANT_BULK_UPDATE_MUTATION, variables
+    )
```

Comment:
Because of using the `mocked_save`, in fact we skip the `save_variants` function execution. In order to mimic concurrency I suggest to use `before_after` and do this:
```
    with before_after.before(
        "saleor.graphql.product.bulk_mutations.product_variant_bulk_update.ProductVariantBulkUpdate.save_variants",
        add_stock_before_save
    ):
        response = staff_api_client.post_graphql(
            PRODUCT_VARIANT_BULK_UPDATE_MUTATION, variables
        )
```

Reply:
Unfortunately `before_after` is currently bugged with `classmethod`s. So I will leave this as it is.